### PR TITLE
error for invalid media type should not be completely swallowed

### PR DESCRIPTION
**Pull request by aristapimenta**
Hi all!

I'm developing a `/metrics` endpoint for a proprietary product and found that my server was sending this header in the response:

```
Content-Type: application/openmetrics-text; version=1.0.0; charset=utf-8; charset=UTF-8
```

As a result, `mime.ParseMediaType(contentType)` will return `errors.New("mime: duplicate parameter name")` because of the `charset` that appears twice (this is a bug that I will fix on my side, of course).

However, `textparse.New()` swallows this error and just fallbacks to `NewPromParser()` (when I wanted it to return `NewOpenMetricsParser()`), which is okay but not great, since it was hard for me to find this bug.

So I thought this debug-level log could be helpful to other people. Cheers!

### CODE REVIEW

**Comment by roidelapluie** [comment]
I feel like this extra call on every scrape might be too much, but we could expose somehow that we parse using openmetric or text. However I am unsure about how we would store/display that.

**Comment by aristapimenta** [comment]
> I feel like this extra call on every scrape might be too much, but we could expose somehow that we parse using openmetric or text. However I am unsure about how we would store/display that.

it's possible to pass down the logger from `*scrapeLoop` to `textparse.New()` in order to avoid the extra call to `mime.ParseMediaType()`. is that what you mean? (aside from tests, this is the only call to `textparse.New()`)

**Comment by aristapimenta** [comment]
(sorry for the push-forces, I'm getting used to the `Signed-off-by` CI check)

**Comment by roidelapluie** [comment]
I would keep the logger where it is and let the function return the error, WDYT?

**Comment by aristapimenta** [comment]
> I would keep the logger where it is and let the function return the error, WDYT?

also works, but it's weird when a function returns an error alongside a valid result 🤔 but it's your call of course

**Comment by aristapimenta** [comment]
> > I would keep the logger where it is and let the function return the error, WDYT?
> 
> also works, but it's weird when a function returns an error alongside a valid result 🤔 but it's your call of course
> 
> (but I agree that passing down this logger feels a bit off)

unless the caller decides by itself to create the default value, which sounds better 💡

**Comment by roidelapluie** [comment]
We could call that return "warning" instead of error.

File: model/textparse/interface.go [hunk 1]
```diff
@@ -61,12 +61,23 @@ type Parser interface {
...
+func New(b []byte, contentType string) (Parser, error) {
+	if contentType == "" {
+		return NewPromParser(b), nil
+	}
+
 	mediaType, _, err := mime.ParseMediaType(contentType)
-	if err == nil && mediaType == "application/openmetrics-text" {
-		return NewOpenMetricsParser(b)
+	if err != nil {
+		return nil, err
```
Comment (roidelapluie) [inline model/textparse/interface.go:74]: This is a breaking change, it would not be acceptable.

File: model/textparse/interface.go [hunk 2]
```diff
@@ -61,12 +61,23 @@ type Parser interface {
 }
 
 // New returns a new parser of the byte slice.
-func New(b []byte, contentType string) Parser {
+//
+// If content type is empty (NewPromParser(b), nil) is returned.
+// Any other invalid content type will return (nil, err).
```
Comment (roidelapluie) [inline model/textparse/interface.go:66]:
```suggestion
// This function always returns a valid parser, but might additionally return an error if the content type can not be parsed.
```

File: model/textparse/interface.go (hunk 1 above)
Comment (aristapimenta) [reply model/textparse/interface.go]: fixed, thanks

File: model/textparse/interface.go (hunk 2 above)
Comment (aristapimenta) [reply model/textparse/interface.go]: fixed, thanks

**Comment by aristapimenta** [comment]
@roidelapluie fixed the breaking change, thanks!

I also thought that was sensible to raise the log level from debug to warning, to give more visibility. Not sure if this could also be too much on every scrape and might flood deployments out there, so please let me know if I need to change it back, but this looks like something that people might want to fix. Actually, the scrape wouldn't work at all in a case like mine, where the response payload is consistent with the "semi-correct" `Content-Type` header, i.e. returning OpenMetrics text, but the scrape code fallbacks to the other parser. The error showing up in the web ui is a parsing error, which is scary if you're already using (parts of) a supported library like https://github.com/prometheus/client_java.

File: scrape/scrape.go [hunk 3]
```diff
@@ -1418,8 +1418,16 @@ type appendErrors struct {
 }
 
 func (sl *scrapeLoop) append(app storage.Appender, b []byte, contentType string, ts time.Time) (total, added, seriesAdded int, err error) {
+	p, warning := textparse.New(b, contentType)
+	if warning != nil {
+		level.Warn(sl.l).Log(
```
Comment (roidelapluie) [inline scrape/scrape.go:1423]: Even if it is warning, this should be logged as debug, because it would be too verbose.

File: promql/fuzz.go [hunk 4]
```diff
@@ -57,7 +57,13 @@ const (
 )
 
 func fuzzParseMetricWithContentType(in []byte, contentType string) int {
-	p := textparse.New(in, contentType)
+	p, warning := textparse.New(in, contentType)
+	if warning != nil {
+		// an invalid content type is being passed, which should not happen
```
Comment (roidelapluie) [inline promql/fuzz.go:62]:
```suggestion
		// An invalid content type is being passed, which should not happen
```

File: promql/fuzz.go [hunk 5]
```diff
@@ -57,7 +57,13 @@ const (
...
+		// in this context
```
Comment (roidelapluie) [inline promql/fuzz.go:63]:
```suggestion
		// in this context.
```

File: scrape/scrape.go (hunk 3 above)
Comment (aristapimenta) [reply scrape/scrape.go]: fixed, thanks

File: promql/fuzz.go (hunk 5 above)
Comment (aristapimenta) [reply promql/fuzz.go]: fixed, thanks

File: model/textparse/interface_test.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+package textparse
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestNew(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (aristapimenta) [reply model/textparse/interface_test.go]: fixed, thanks

File: model/textparse/interface_test.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+			warning:        "",
+		},
+		"some-other-valid-content-type": {
+			contentType:    "text/html",
+			validateParser: requirePromParser,
+			warning:        "",
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			tt := tt // copy to local var before going parallel
```
Comment (roidelapluie) [inline model/textparse/interface_test.go:99]:
```suggestion
			tt := tt // Copy to local variable before going parallel.
```

**Review by roidelapluie — COMMENTED** [review]
Just a last nit.

File: model/textparse/interface_test.go (hunk 7 above)
Comment (aristapimenta) [reply model/textparse/interface_test.go]: fixed, thanks

**Comment by roidelapluie** [comment]
Close/reopen to get CI running.

**Review by roidelapluie — APPROVED** [review]

**Comment by roidelapluie** [comment]
Thanks! After double thinking, it sounds better to call it err everywhere, I have open #10267 as followup.

**Pull request merged**